packages/gatsby/src/utils/worker/render-html.ts

```diff
@@ -1,50 +1,315 @@
+  inFlightResourcesForTemplate.delete(pageData.componentChunkName)
+
+  return resources
+}
+
+export const renderHTMLProd = async ({
   htmlComponentRendererPath,
   paths,
-  stage,
   envVars,
+  timestamp,
 }: {
   htmlComponentRendererPath: string
   paths: Array<string>
-  stage: Stage
   envVars: Array<Array<string>>
+  timestamp: number
 }): Promise<Array<unknown>> => {
-  // This is being executed in child process, so we need to set some vars
-  // for modules that aren't bundled by webpack.
-  envVars.forEach(([key, value]) => (process.env[key] = value))
-
-  const htmlComponentRenderer = require(htmlComponentRendererPath)
-
-  return Promise.map(
-    paths,
-    path =>
-      new Promise((resolve, reject) => {
-        try {
-          htmlComponentRenderer.default(path, (_throwAway, htmlString) => {
-            if (stage === `develop-html`) {
-              return resolve(htmlString)
-            } else {
-              return resolve(
-                fs.outputFile(
-                  getPageHtmlFilePath(join(process.cwd(), `public`), path),
-                  htmlString
-                )
-              )
-            }
-          })
-        } catch (e) {
-          // add some context to error so we can display more helpful message
-          e.context = {
-            path,
-          }
-          reject(e)
-        }
+  const publicDir = join(process.cwd(), `public`)
+
+  if (timestamp !== lastRenderHTMLCallTimestamp) {
+    staticQueryResultCache.clear()
+    resourcesForTemplateCache.clear()
+    // This is being executed in child process, so we need to set some vars
+    // for modules that aren't bundled by webpack.
+    envVars.forEach(([key, value]) => (process.env[key] = value))
+
+    htmlComponentRenderer = require(htmlComponentRendererPath)
+
+    webpackStats = await readWebpackStats(publicDir)
+
+    lastRenderHTMLCallTimestamp = timestamp
+  }
+
+  return Bluebird.map(paths, async pagePath => {
+    try {
+      const pageData = await readPageData(publicDir, pagePath)
+      const resourcesForTemplate = await getResourcesForTemplate(pageData)
+
+      const htmlString = htmlComponentRenderer.default({
+        pagePath,
+        pageData,
+        ...resourcesForTemplate,
       })
-  )
+
+      return fs.outputFile(getPageHtmlFilePath(publicDir, pagePath), htmlString)
+    } catch (e) {
+      // add some context to error so we can display more helpful message
+      e.context = {
+        path: pagePath,
+      }
+      throw e
+    }
+  })
+}
+
+// TODO: remove when DEV_SSR is done
+export const renderHTMLDev = async ({
+  htmlComponentRendererPath,
+  paths,
+  envVars,
+  timestamp,
+}: {
+  htmlComponentRendererPath: string
+  paths: Array<string>
+  envVars: Array<Array<string>>
+  timestamp: number
+}): Promise<Array<unknown>> => {
+  const outputDir = join(process.cwd(), `.cache`, `develop-html`)
+  if (timestamp !== lastRenderHTMLCallTimestamp) {
```

Comment:
We don't really about the time(stamp) per se. Rather about whatever we do inside this conditional is done once per `renderHTMLQueue` call in each worker process of `jest-worker` ( https://github.com/gatsbyjs/gatsby/blob/2fb0aa10771c62ca945277b279b4ba015fcac046/packages/gatsby/src/commands/build-html.ts#L132-L168 ). So this could be some kind of counter, uuid v4 etc

I thought of maybe having separate method for "setting up" worker process that is called for each worker before calling `renderHTMLProd` with page chunks, but `jest-worker` doesn't seem to easily allow to call some arbitrary function in each worker once those are already alive. `jest-worker` does have special `setup` method ( https://www.npmjs.com/package/jest-worker#setting-up-and-tearing-down-the-child-process ) - but this is one time called for entire life of each jest-worker child and doesn't seem like it can be arbitrarily called at any point

Reply:
Yeah, maybe `sessionId` or something like this. My point is only about `timestamp` name being a bit confusing for this usage.

Reply:
renamed and added some comments about `sessionId` in https://github.com/gatsbyjs/gatsby/pull/29396/commits/3bac1757563f923dda48872542237e03068d739d